packages/consumption/src/modules/requests/utility/validateRelationshipAttributesWithinRequest.ts

```diff
@@ -0,0 +1,206 @@
+import {
+    CreateAttributeRequestItem,
+    ProposeAttributeRequestItem,
+    ReadAttributeRequestItem,
+    RelationshipAttribute,
+    RelationshipAttributeQuery,
+    RequestItem,
+    RequestItemGroup
+} from "@nmshd/content";
+import { CoreAddress } from "@nmshd/core-types";
+import { ConsumptionCoreErrors } from "../../../consumption/ConsumptionCoreErrors";
+import { ValidationResult } from "../../common";
+import { DecideRequestItemGroupParametersJSON } from "../incoming/decide/DecideRequestItemGroupParameters";
+import { DecideRequestItemParametersJSON } from "../incoming/decide/DecideRequestItemParameters";
+
+interface RelationshipAttributeFragment {
+    owner: string;
+    key: string;
+    value: { "@type": string };
+}
+
+type ContainsDuplicateRelationshipAttributeFragmentsResponse =
+    | { containsDuplicates: false }
+    | {
+          containsDuplicates: true;
+          duplicateFragment: RelationshipAttributeFragment;
+      };
+
+export function validateKeyUniquenessOfRelationshipAttributesWithinOutgoingRequest(items: (RequestItem | RequestItemGroup)[], recipient?: CoreAddress): ValidationResult {
+    const fragmentsOfMustBeAcceptedItemsOfRequest: RelationshipAttributeFragment[] = [];
+    for (const item of items) {
+        if (item instanceof RequestItemGroup) {
+            const fragmentsOfMustBeAcceptedItemsOfGroup = extractRelationshipAttributeFragmentsFromMustBeAcceptedItemsOfGroup(item, recipient);
+            if (fragmentsOfMustBeAcceptedItemsOfGroup) fragmentsOfMustBeAcceptedItemsOfRequest.push(...fragmentsOfMustBeAcceptedItemsOfGroup);
+        } else {
+            const fragmentOfMustBeAcceptedRequestItem = extractRelationshipAttributeFragmentFromMustBeAcceptedRequestItem(item, recipient);
+            if (fragmentOfMustBeAcceptedRequestItem) fragmentsOfMustBeAcceptedItemsOfRequest.push(fragmentOfMustBeAcceptedRequestItem);
+        }
+    }
```

Comment:
I think you can simply call `extractRelationshipAttributeFragmentsFromMustBeAcceptedItemsOfGroup` here, since you already implemented the recursion there. Probably some names need to be adjusted then.

Reply:
Also you are not using forEach in instances where it could be used

Reply:
> I think you can simply call `extractRelationshipAttributeFragmentsFromMustBeAcceptedItemsOfGroup` here, since you already implemented the recursion there. Probably some names need to be adjusted then.

At first glance, this approach does not work with meaningful names or with the appropriate typing of the input parameters, but I'll see if I can get rid of the redundant recursion.

> Also you are not using forEach in instances where it could be used

We use `items.forEach((item) => { ... } )` compared to `for (const item of items) { ... }` less often in our codebase, which is why I would stick with the second variant when refactoring.

Reply:
> At first glance, this approach does not work with meaningful names or with the appropriate typing of the input parameters, but I'll see if I can get rid of the redundant recursion.

Again, the not yet well thought-out features of the nested RequestItemGroups play a role, which is why the type of the `items` property of the Request and the `items` property of the RequestItemGroup do not match, which is why I could not implement your suggestion. Again, the not yet well thought-out features of the nested RequestItemGroups play a role, which is why the type of the `items` property of the request and the `items` property of the RequestItemGroup do not match, which is why I could not implement your suggestion.

Reply:
I'll take another look and get back to you. :)